### ISSUE TITLE
[@types/yargs] .middleware() accepts async functions as well, but the typings don't show this

### ISSUE THREAD

**Issue by fabis94** [DefinitelyTyped/DefinitelyTyped#52060]
The middleware() function from yargs also accepts async middlewares, but the typing only allows for synchronous ones.

![image](https://user-images.githubusercontent.com/938316/112860414-779d9700-90bc-11eb-9a65-f3e9b1f732c4.png)

Here's the doc: https://yargs.js.org/docs/#api-reference-middlewarecallbacks-applybeforevalidation

This happens with yargs 16.2.0 and @types/yargs 16.0.1

Definitions by: @poelstra @mizunashi-mana @pushplay @JimiC @steffenvv @forivall @ExE-Boss @Aankhen

**Comment by fabis94** [comment]
Here's the relevant bit from the actual JS Build of yargs:
![image](https://user-images.githubusercontent.com/938316/112861122-30fc6c80-90bd-11eb-8454-a9badfa3fbcc.png)

**Comment by forivall** [comment]
It's still possible to pass an async function to `.middleware` without any typescript error.

```ts
type VoidFn = () => void;

function takesVoidFn (asdf: VoidFn) {
  console.log(asdf)
}

takesVoidFn(async () => {})
```
[Playground Link](https://www.typescriptlang.org/play?jsx=0&ts=4.2.3#code/FAFwngDgpgBAagewJYBMBiA7GBeGAKAShwD4YA3ZFAbmGADMBXDAYxCQSxAEMBrKAZ0SpM+LvxR0AXPEqYiAb2AwYzDvwQAbKADoNCAOZ4xEgsAC+tbn0GyMR-mBb4i2UvLOmgA)

**Comment by fabis94** [comment]
You're right, I'm sorry! It turned out to be an ESLint error - "Promise returned in function argument where a void return was expected.eslint@typescript-eslint/no-misused-promises".

It'd still be nice if the typings would correctly allow for async functions so that I wouldn't have to ignore this ESLint rule, but it does work outside of ESlint, true

**Comment by forivall** [comment]
Ah! I also discovered that `void | undefined` does make a strict void returning function that `tsc` checks. I'll wait for other maintainers' input on this and my draft PR.
